### DISCUSSION TITLE
feature request - block merge until all reviewers have approved

### DISCUSSION THREAD

**Discussion by osievert-gpfw** [community/community#55218] [Pull Requests]
### Select Topic Area

Product Feedback

## Problem Statement

When working with github in a team setting, a lot of pull requests have a single reviewer, and as soon as that reviewer approves we merge the pull request. That makes sense. But when multiple reviewers have been selected, we sometimes find that pull requests are merged after a single reviewer has approved but another reviewer is in the middle of reviewing.

This is mostly a coordination problem, but sometimes tooling is the best way to coordinate.

Branch protection rules today allow pull requests to have a sepcified number of approvals before being merged. I currently use that feature, and have that set to "1" to ensure a review has happened. I could use this feature to ensure a second (or third, or fourth) reviewer gets a chance to complete their review, but then every pull request would require two/three/four reviews, and we have a lot of pull requests that only really need one review.

## Feature Request

Would it be possible to have another branch protection setting that is similar to this: block merging until all reviewers have approved?

## Benefit

With this new requested feature, if a pull request lists one reviewer, then only that one approval is needed in order to merge. If two reviewers were identified, both would have to approve before the pull request could be merged. This allows simple pull requests to stay simple, but ensures that more complicated pull requests get reviews from all the people who have expertise about the change (or who are impacted by the change).

## Commentary

I could see this new feature and the existing "block merging until N approvals have been given" feature interacting in a complementary way, for example:
- if no reviewers have been specified, and the minimum number of approvals is zero, then allow merging without any reviews
- if no reviewers have been specified, and the minimum number of approvals is one, then require one approval to merge (could be anyone with admin or write privilges in the repo)
- if two reviewers were specified (one of which does not have admin or write privileges), and the minimum number of approvals is three, then both specified reviewers need to approve, plus one additional approval from anyone with admin or write privileges, in order to merge

## References

This feature request is similar to, but not exactly the same, as what is discussed in  https://github.com/orgs/community/discussions/5289

**Comment by booiiing** [comment]
I believe [discussion#48162](https://github.com/orgs/community/discussions/48162) is about the same?
really love your detailed description, @osievert-gpfw, but since the other one is older and has more participation, would you mind sharing it there?

@everyone who upvoted/subscribed here: please feel encouraged to do the same on the other idea in order to push it to github product owners :)

**Reply by osievert-gpfw** [reply]
Done - thanks for linking these two identical requests together.

**Comment by AtzeDeVries** [comment]
you can use https://github.com/palantir/policy-bot to achieve this
